model_building/model_building.py

```diff
@@ -1,254 +1,265 @@
-        self._logger.info("Best models:")
-        self._logger.info("-->")
-        for tec, conf in best_confs.items():
-            self._logger.info("%s:", tec)
-            self._logger.info("---" + conf.print_model())
-        self._logger.info("<--")
-        self._logger.removeHandler(file_handler)
-        file_handler.close()
-
-        # Return the regressor
-        return best_regressors[best_technique]
+"""
+Copyright 2019 Marco Lattuada
+Copyright 2021 Bruno Guindani
+Copyright 2022 Nahuel Coliva
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import logging
+import multiprocessing
+import os
+import pickle
+import random
+import tqdm
+
+import custom_logger
+import data_preparation.normalization
+import model_building.experiment_configuration as ec
+import model_building.generators_factory as gf
+import regressor
+import results as re
+
+
+class ModelBuilding:
+    """
+    Entry point of the model building phase, i.e., where all the regressions are actually performed
+
+    The process method do the following steps:
+        - Create the generators through the factory
+        - Build the model for each ExperimentConfiguration
+        - Evaluate the MAPE on different sets of each ExperimentConfiguration
+        - Identify the best regressor of each technique
+        - Retrain the best regressors with the whole dataset
+        - Dump the best regressors in pickle format
+
+    Attributes
+    ----------
+    random_generator : Random
+        The internal random generator
+
+    _logger: Logger
+        The logger used by this class
+
+    Methods
+    ------
+    _process_wrapper()
+        Wrapper used internally for parallel execution of experiments
+    process()
+        Generates the set of expriment configurations to be evaluated
+    """
+
+    def __init__(self, seed):
+        """
+        Parameters
+        ----------
+        seed: integer
+            The seed to be used for the internal random generator
+        """
+        self._random_generator = random.Random(seed)
+        self._logger = custom_logger.getLogger(__name__)
+
+    def _process_wrapper(self, experiment_configuration):
+        """
+        Wrapper used internally for parallel execution of experiments
+        """
+        if self.debug:
+            # Do not use try-except mechanism
+            experiment_configuration.train(parallelism=True)
+        else:
+            try:
+                experiment_configuration.train(parallelism=True)

```

Comment:
Shouldn't this be `parallelism=self._parallelism`? It's unclear to me when the `_parallelism` flag comes up. It should work such that regressor-level parallelism is enabled when experiments-level parallelism is not, and viceversa

Reply:
self.parallelism is inizialized during training (first line of the method in experiment_configuration.py) and used only in xgboost_experiment_configuration.py to set the correct nthread value

Reply:
Alright, now I get it. I was also missing the fact that this function is only used for *parallel* execution of experiments.

That being said, `parallelism` is a terrible name for this flag, because I thought it would activate XGBoost parallelism, not disable it 😆 I would change it to something like `disable_model_parallelism` and add documentation for it in the appropriate function

Reply:
The rationale was "it signals whether the parallelism has been enabled or not from the conf file", anyways i changed the name and documented it